code/modules/bitrunning/virtual_domain/modular_mob_segment.dm

```diff
@@ -0,0 +1,180 @@
+#define SPAWN_ALWAYS 100
+#define SPAWN_LIKELY 85
+#define SPAWN_UNLIKELY 35
+#define SPAWN_RARE 10
+
+/datum/modular_mob_segment
+	/// Set this to false if you want explicitly what's in the list to spawn
+	var/randomized = TRUE
+	/// Spawn a random amount from the list
+	var/pick_random_of = 0
+	/// The list of mobs to spawn
+	var/list/mob/living/mobs = list()
+	/// The mobs spawned from this segment
+	var/list/spawned_mob_refs = list()
+	/// Chance this will spawn (1 - 100)
+	var/probability = SPAWN_LIKELY
+
+/// Spawns mobs in a circle around the location
+/datum/modular_mob_segment/proc/spawn_mobs(turf/origin)
+	if(!prob(probability))
+		return
+
+	var/current_index = 1
+
+	var/total_amount
+	if(pick_random_of > 0)
+		total_amount = pick_random_of
+	else
+		if(!randomized)
+			total_amount = length(mobs)
+		else
+			total_amount = rand(1, 6)
+
+	shuffle_inplace(mobs)
+
+	var/turf/nearby = view(5, origin)
+
+	var/spawned = FALSE
+	for(var/index in 1 to total_amount)
+		spawned = FALSE
+
+		while(!spawned)
+			if(length(nearby) == 0)
+				break
+
+			for(var/turf/open/possible_turf in nearby)
+				if(possible_turf.is_blocked_turf())
+					nearby -= possible_turf
+					continue
+
+				var/path
+				if(randomized)
+					path = pick(mobs)
+				else
+					path = mobs[current_index]
+					current_index++
+
+				var/mob/living/mob = new path(possible_turf)
+				nearby -= possible_turf
+				spawned = TRUE
+				spawned_mob_refs.Add(WEAKREF(mob))
+
+				break
```

Comment:
This is likely causing infinite loops when `for(var/turf/open/possible_turf in nearby)` never finds a viable turf.